# HW3

### Emmanuella Eguche

## Q1

In [431]:
from pandas import Series, DataFrame
import pandas as pd
%pylab inline


Populating the interactive namespace from numpy and matplotlib


D:\Anaconda\lib\site-packages\IPython\core\magics\pylab.py:160: UserWarning: pylab import has clobbered these variables: ['quantile']
`%matplotlib` prevents importing * from pylab and numpy
  "\n`%matplotlib` prevents importing * from pylab and numpy"


In [432]:
# read in gold.txt
gold = pd.read_csv('./gold.txt', sep=" ", header=None)
gold[['url','category']] = gold[0].str.split(expand=True)
gold = gold[['url','category']]
gold

url category
0                  http://0800-horoscope.com        G
1                         http://18games.net        X
2                       http://1pixelout.net        G
3     http://1porno.com/004/teens/index.html        X
4      http://1stpussy.com/005/16/index.html        X
...                                      ...      ...
1512                     http://zipcodez.com        G
1513                          http://zml.com        P
1514                     http://zonezero.com        G
1515                       http://zooped.com        R
1516                        http://ztrua.com        G

[1517 rows x 2 columns]

In [433]:
# read in labels.txt
labels = pd.read_csv('./labels.txt', sep=" ", header=None)
labels[['turk','url','category']] = labels[0].str.split(expand=True)
labels = labels[['turk','url','category']]
labels

turk               url category
0      A1OT3A29R9N1DG     http://000.cc        P
1      A1PXXEOGQ76RNJ     http://000.cc        G
2      A1PXXEOGQ76RNJ     http://000.cc        G
3      A21US576U8SCO4     http://000.cc        G
4      A2LGX47NN7C5D3     http://000.cc        G
...               ...               ...      ...
92716   AFHMR93GDJCGT  http://zz868.com        G
92717  A13F119AMOA54H     http://zzx.cc        X
92718  A1S0C0EB7A6FFM     http://zzx.cc        P
92719  A21US576U8SCO4     http://zzx.cc        G
92720  A29T6EIMK6R7O0     http://zzx.cc        X

[92721 rows x 3 columns]

# Q2

In [434]:
# create the labels_on_gold dataframe
labels_on_gold = labels.merge(gold, on='url', suffixes = ['_label','_gold'])
labels_on_gold  


turk                        url category_label category_gold
0     A1253FXHCZ9CWM  http://0800-horoscope.com              G             G
1     A153PKAL7OAY36  http://0800-horoscope.com              G             G
2     A1FV9SAPL5C6KY  http://0800-horoscope.com              G             G
3     A1JTOT0DWM6QGL  http://0800-horoscope.com              G             G
4     A1PXXEOGQ76RNJ  http://0800-horoscope.com              G             G
...              ...                        ...            ...           ...
3319  A35CBDDGBM7T6E       http://yousendit.com              G             G
3320  A3J86MK3VIE6ST       http://yousendit.com              G             G
3321   A4GP97LAQCCHJ       http://yousendit.com              G             G
3322   AAQNUEUU7H1Y5       http://yousendit.com              G             G
3323   ANC9EWZTGN5HK       http://yousendit.com              G             G

[3324 rows x 4 columns]

In [435]:
def url_g(s):
    if s in gold['url'].values:
        return True
    else:
        return False
    
labels_unknown = labels[labels['url'].apply(url_g) == False]
labels_unknown

turk               url category
0      A1OT3A29R9N1DG     http://000.cc        P
1      A1PXXEOGQ76RNJ     http://000.cc        G
2      A1PXXEOGQ76RNJ     http://000.cc        G
3      A21US576U8SCO4     http://000.cc        G
4      A2LGX47NN7C5D3     http://000.cc        G
...               ...               ...      ...
92716   AFHMR93GDJCGT  http://zz868.com        G
92717  A13F119AMOA54H     http://zzx.cc        X
92718  A1S0C0EB7A6FFM     http://zzx.cc        P
92719  A21US576U8SCO4     http://zzx.cc        G
92720  A29T6EIMK6R7O0     http://zzx.cc        X

[89397 rows x 3 columns]

## Q3

In [436]:
# groupby function
grps = labels_on_gold.groupby('turk')

# build a dataframe rater_goodness
rater_goodness = pd.DataFrame()
rater_goodness['number_of_ratings'] = labels_on_gold.groupby('turk').count()['url']

for key, group in grps:
    correctness = len(group[group['category_label'] == group['category_gold']])/len(group) 
    rater_goodness.loc[key,'avg_correctness'] = correctness
    
rater_goodness

number_of_ratings  avg_correctness
turk                                              
A112DVP1KG4QZU                  1         1.000000
A1253FXHCZ9CWM                 29         0.517241
A12CY1Q7XKJJDE                  1         1.000000
A12RE8G66WTO8B                 20         0.750000
A12Y1GTGIQDGRA                  3         0.333333
...                           ...              ...
AYWOBMTEGFVVI                   1         0.000000
AZ1S5BIG5R6P6                  35         0.714286
AZSDW7JJL5VA7                 100         0.680000
AZVB3RZW0OR1H                   3         0.666667
amurray                        15         0.933333

[269 rows x 2 columns]

## Q4

In [437]:
c = rater_goodness['avg_correctness']
rater_goodness['odds'] = c/(1.001-c)
rater_goodness

number_of_ratings  avg_correctness         odds
turk                                                           
A112DVP1KG4QZU                  1         1.000000  1000.000000
A1253FXHCZ9CWM                 29         0.517241     1.069214
A12CY1Q7XKJJDE                  1         1.000000  1000.000000
A12RE8G66WTO8B                 20         0.750000     2.988048
A12Y1GTGIQDGRA                  3         0.333333     0.499251
...                           ...              ...          ...
AYWOBMTEGFVVI                   1         0.000000     0.000000
AZ1S5BIG5R6P6                  35         0.714286     2.491281
AZSDW7JJL5VA7                 100         0.680000     2.118380
AZVB3RZW0OR1H                   3         0.666667     1.994018
amurray                        15         0.933333    13.793103

[269 rows x 3 columns]

## Q5

In [438]:
rater_goodness[rater_goodness['number_of_ratings'] >= 20].sort_values\
(by = "avg_correctness", ascending = False).head(10)

number_of_ratings  avg_correctness       odds
turk                                                         
A2U0R4X38GUKZE                 20         0.950000  18.627451
A22C0PJUBFJTI0                 36         0.916667  10.869565
A23YQUBXZPKILZ                 24         0.875000   6.944444
ATVALOQVDCMZW                 103         0.854369   5.826657
A1HIXWH4OXT8S4                 40         0.825000   4.687500
A3220HG1O83HQ4                 22         0.818182   4.475385
A32W20KGQXS0LL                 25         0.800000   3.980100
A20PWAB7G3HDHU                 20         0.800000   3.980100
AJSJVK40F5HM6                  28         0.785714   3.649635
A31OCN4MNHUQ6W                184         0.777174   3.472222

## Q6

In [439]:
# create a correlation matrix between the number and correctness
rater_goodness[['number_of_ratings','avg_correctness']].corr()

# There seems to be not much relatedness between the two variables.
# We will further run OLS regression to testify this result.

number_of_ratings  avg_correctness
number_of_ratings           1.000000         0.045292
avg_correctness             0.045292         1.000000

In [440]:
# create a linear model and test accuracy
import statsmodels.api as sm
from patsy import dmatrices

y, X = dmatrices('number_of_ratings ~ avg_correctness', data=rater_goodness, return_type='dataframe')
model = sm.OLS(y, X)       # Set up the model
result = model.fit()       # Fit model (find the intercept and slopes)
print(result.summary())

# The 95% confidence interval of the coefficient of independent variable includes zero
# and is not statistically significant.
# There is no proved significant relationship between number and correctness

                            OLS Regression Results                            
Dep. Variable:      number_of_ratings   R-squared:                       0.002
Model:                            OLS   Adj. R-squared:                 -0.002
Method:                 Least Squares   F-statistic:                    0.5488
Date:                Thu, 06 Aug 2020   Prob (F-statistic):              0.459
Time:                        13:44:30   Log-Likelihood:                -1232.3
No. Observations:                 269   AIC:                             2469.
Df Residuals:                     267   BIC:                             2476.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept          10.3317      3.092     

## Q7

In [441]:
# calculate the first 25 percentile
quantile = rater_goodness['number_of_ratings'].quantile(.75)
rater_sort = rater_goodness.sort_values(by = 'number_of_ratings', ascending = False)
rater_q = rater_sort[rater_sort['number_of_ratings'] >= quantile]
rater_q

number_of_ratings  avg_correctness      odds
turk                                                        
A31OCN4MNHUQ6W                184         0.777174  3.472222
A153PKAL7OAY36                148         0.722973  2.600369
A2BTR0GQ5B5JI6                126         0.523810  1.097695
ATVALOQVDCMZW                 103         0.854369  5.826657
AZSDW7JJL5VA7                 100         0.680000  2.118380
...                           ...              ...       ...
AV30XWLULCP43                  13         0.615385  1.595851
A1IB9WML70CU89                 12         0.833333  4.970179
A4GP97LAQCCHJ                  12         0.583333  1.396648
ARQCISK0K4ZGA                  11         0.545455  1.197366
A35HNSSK5MHRAQ                 11         0.636364  1.745201

[68 rows x 3 columns]

In [442]:
def url_r(s):
    if s in rater_q.index:
        return True
    else:
        return False
    
unknown = labels_unknown[labels_unknown['turk'].apply(url_r) == True]
unknown['odds'] = rater_q.loc[unknown['turk']]['odds'].values
unknown

D:\Anaconda\lib\site-packages\ipykernel_launcher.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


turk               url category      odds
1      A1PXXEOGQ76RNJ     http://000.cc        G  1.047852
2      A1PXXEOGQ76RNJ     http://000.cc        G  1.047852
3      A21US576U8SCO4     http://000.cc        G  1.330229
10     A1PXXEOGQ76RNJ    http://0000.jp        G  1.047852
11     A1XFA0A9Z3LAZJ    http://0000.jp        G  3.732272
...               ...               ...      ...       ...
92707   A4GP97LAQCCHJ  http://zynga.com        G  1.396648
92709  A1PXXEOGQ76RNJ  http://zz868.com        G  1.047852
92715   A4AD9ZMVU9SVA  http://zz868.com        P  1.851852
92719  A21US576U8SCO4     http://zzx.cc        G  1.330229
92720  A29T6EIMK6R7O0     http://zzx.cc        X  1.620746

[41647 rows x 4 columns]

In [443]:
overall_odds = unknown.groupby(['url','category'])[['odds']].agg('product')
overall_odds

odds
url                         category           
http://0-101.net            G          2.155963
http://000.cc               G          1.460583
http://0000.jp              G         14.488244
http://000relationships.com G          9.914590
                            P          1.851852
...                                         ...
http://zynga.com            G          3.704611
http://zz868.com            G          1.047852
                            P          1.851852
http://zzx.cc               G          1.330229
                            X          1.620746

[14114 rows x 1 columns]

## Q8

In [444]:
# pivot the overall_odds table
pivot = pd.pivot_table(overall_odds,
                          index = 'url',
                          columns=['category'],
                          values = ['odds'],
                          aggfunc = max)

# transpose and find the top category
category_top = DataFrame(pivot['odds'].T.idxmax())
category_top.head(5)

0
url                           
http://0-101.net             G
http://000.cc                G
http://0000.jp               G
http://000relationships.com  G
http://000vitamins.com       G

In [445]:
# find the max odds
odds_top = DataFrame(pivot['odds'].T.max())
odds_top.head(5)

0
url                                   
http://0-101.net              2.155963
http://000.cc                 1.460583
http://0000.jp               14.488244
http://000relationships.com   9.914590
http://000vitamins.com        3.784982

In [446]:
# merge dataframes
result_75 = category_top.merge(odds_top, left_on = 'url', right_on = 'url')

# change column names
result_75.columns = ['top category', 'top odds']

result_75

top category   top odds
url                                                
http://0-101.net                       G   2.155963
http://000.cc                          G   1.460583
http://0000.jp                         G  14.488244
http://000relationships.com            G   9.914590
http://000vitamins.com                 G   3.784982
...                                  ...        ...
http://zwinky.com                      G  16.141443
http://zylom.com                       G   1.735526
http://zynga.com                       G   3.704611
http://zz868.com                       P   1.851852
http://zzx.cc                          X   1.620746

[10610 rows x 2 columns]

## Q9

In [447]:
# repeat Q7 for first 75%
# calculate the first 75 percentile
quantile_25 = rater_goodness['number_of_ratings'].quantile(.25)
rater_sort25 = rater_goodness.sort_values(by = 'number_of_ratings', ascending = False)
rater_q25 = rater_sort25[rater_sort25['number_of_ratings'] >= quantile_25]

def url_r(s):
    if s in rater_q25.index:
        return True
    else:
        return False
    
unknown25 = labels_unknown[labels_unknown['turk'].apply(url_r) == True]
unknown25['odds'] = rater_q25.loc[unknown25['turk']]['odds'].values

overall_odds25 = unknown25.groupby(['url','category'])[['odds']].agg('product')
overall_odds25

D:\Anaconda\lib\site-packages\ipykernel_launcher.py:14: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


odds
url                         category              
http://0-101.net            G         2.155963e+00
http://000.cc               G         2.181050e+03
                            P         9.980040e-01
http://0000.jp              G         2.877556e+07
http://000relationships.com G         0.000000e+00
...                                            ...
http://zz868.com            G         1.209566e+16
                            P         1.851852e+00
http://zzx.cc               G         1.330229e+00
                            P         9.980040e-01
                            X         1.620746e+00

[17794 rows x 1 columns]

In [448]:
# Repeat Q8 with the first 75%
# pivot the overall_odds25 table
pivot25 = pd.pivot_table(overall_odds25,
                          index = 'url',
                          columns=['category'],
                          values = ['odds'],
                          aggfunc = max)

# transpose and find the top category
category_top25 = DataFrame(pivot25['odds'].T.idxmax())

# find the max odds
odds_top25 = DataFrame(pivot25['odds'].T.max())

# merge dataframes
result_25 = category_top25.merge(odds_top25, left_on = 'url', right_on = 'url')

# change column names
result_25.columns = ['top category', 'top odds']

result_25

top category      top odds
url                                                   
http://0-101.net                       G  2.155963e+00
http://000.cc                          G  2.181050e+03
http://0000.jp                         G  2.877556e+07
http://000relationships.com            P  1.851852e+00
http://000vitamins.com                 G  5.014149e+03
...                                  ...           ...
http://zwinky.com                      P  1.000000e+03
http://zylom.com                       G  2.873345e+06
http://zynga.com                       G  5.442125e+01
http://zz868.com                       G  1.209566e+16
http://zzx.cc                          X  1.620746e+00

[10707 rows x 2 columns]

In [449]:
# create the final table by merging the two previous tables
final = result_75.merge(result_25, left_on = 'url',\
                        right_on = 'url',\
                        suffixes = ['_75', '_25'])
final

top category_75  top odds_75 top category_25  \
url                                                                        
http://0-101.net                          G     2.155963               G   
http://000.cc                             G     1.460583               G   
http://0000.jp                            G    14.488244               G   
http://000relationships.com               G     9.914590               P   
http://000vitamins.com                    G     3.784982               G   
...                                     ...          ...             ...   
http://zwinky.com                         G    16.141443               P   
http://zylom.com                          G     1.735526               G   
http://zynga.com                          G     3.704611               G   
http://zz868.com                          P     1.851852               G   
http://zzx.cc                             X     1.620746               X   

                              top odds_25  
url                                        
http://0-101.net             2.155963e+00  
http://000.cc                2.181050e+03  
http://0000.jp               2.877556e+07  
http://000relationships.com  1.851852e+00  
http://000vitamins.com       5.014149e+03  
...                                   ...  
http://zwinky.com            1.000000e+03  
http://zylom.com             2.873345e+06  
http://zynga.com             5.442125e+01  
http://zz868.com             1.209566e+16  
http://zzx.cc                1.620746e+00  

[10610 rows x 4 columns]

In [450]:
compare = pd.crosstab(final['top category_75'], final['top category_25'])
compare

top category_25     G    P    R    X
top category_75                     
G                7911  836  285  237
P                 228  264   65   25
R                  22   36  120   35
X                  70   12   35  429

## The most errors occur when top category_75 predicts G and top category_25 predicts P with 836 occurrences.